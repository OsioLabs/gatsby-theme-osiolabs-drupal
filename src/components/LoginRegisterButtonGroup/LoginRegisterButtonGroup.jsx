import React from 'react';
import { Button } from 'semantic-ui-react';
import DrupalLoginButton from '../drupal-oauth/DrupalLoginButton';

const LoginRegisterButtonGroup = () => (
  <Button.Group>
    <DrupalLoginButton title="Log in" classes="ui button" />
    <Button.Or text="/" />
    <DrupalLoginButton
      title="Sign up"
      useRegistrationLink
      classes="ui button primary"
    />
  </Button.Group>
);

export default LoginRegisterButtonGroup;
