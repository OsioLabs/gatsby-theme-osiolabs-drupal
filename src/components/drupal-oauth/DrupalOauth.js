import retry from 'async-retry';
import qs from 'qs';

class DrupalOauth {
  processing = false;

  /**
   *
   * @param config
   * - drupal_root:
   * - token_url:
   * - authorize_url:
   * - client_id:
   * - client_secret:
   */
  constructor(config) {
    this.config = config;
    this.config.token_url = `${this.config.drupal_root}/oauth/token`;
    this.config.authorize_url = `${this.config.drupal_root}/oauth/authorize`;
  }

  /**
   * Check to see if the current user is logged in.
   *
   * If the user was previously logged in but their access token is expired
   * attempt to retrieve a new token.
   *
   * @param {function} refreshComplete
   *  Optional callback function to call after a refresh token has been
   *  processed.
   *
   * @returns <Mixed>
   *   The current users authorization token, or false.
   */
  isLoggedIn = async refreshComplete =>
    retry(async bail => {
      if (
        typeof localStorage === 'undefined' ||
        typeof window === 'undefined'
      ) {
        // This is being executed outside of the browser, so it's not going to work
        // and we can just bail out now
        bail();
        return false;
      }

      const token =
        localStorage.getItem('drupal-oauth-token') !== null
          ? JSON.parse(localStorage.getItem('drupal-oauth-token'))
          : null;

      // If there is no token, and we're not processing, that means there is no
      // logged in user. So we're done.
      if (token === null && this.processing === false) {
        bail();
        return false;
      }

      // If we're currently trying to retrieve a refersh token from another
      // request then just retry.
      if (this.processing === true) {
        throw new Error('retry');
      }

      // If we've got an active token, the user is logged in, so we return their
      // current token.
      if (
        token !== null &&
        token.expirationDate > Math.floor(Date.now() / 1000)
      ) {
        this.processing = false;
        return token;
      }

      // If there is an existing token, but it's expired, see if we can get a
      // a new one using the refresh token. If not, see if we can get a refresh
      // token.
      this.processing = true;
      await this.getRefreshToken(token).then(fullfilledToken => {
        if (fullfilledToken !== null) {
          this.processing = false;
          if (typeof refreshComplete === 'function') {
            refreshComplete(true);
          }
          return fullfilledToken;
        }

        bail(new Error('Unable to retrieve an access token'));
        return false;
      });

      return false;
    });

  /**
   *
   */
  async handleLogin(username, password, scope) {
    return this.fetchOauthToken(username, password, scope);
  }

  /**
   * Log the current user out.
   *
   * Deletes the token from local storage.
   */
  // eslint-disable-next-line class-methods-use-this
  async handleLogout() {
    return localStorage.removeItem('drupal-oauth-token');
  }

  async getOauthToken(username, password, scope) {
    return this.fetchOauthToken(username, password, scope);
  }

  async getRefreshToken(token) {
    return this.refreshOauthToken(token);
  }

  /**
   * Get an OAuth token from Drupal.
   *
   * Exchange a an authorization code for an OAuth token. Uses ?code and other
   * parameters from the current URL.
   *
   * @returns {Promise<void>}
   *   Returns a promise that resolves with the new token returned from Drupal.
   */
  // eslint-disable-next-line consistent-return
  async fetchOauthToken(url) {
    this.processing = true;

    const queryString = qs.parse(url.split('?')[1], {
      ignoreQueryPrefix: true,
    });

    if (!queryString.code) {
      throw new Error('Can not retrieve authorization token without a code.');
    }

    const formData = new FormData();
    Object.keys(queryString).forEach(item => {
      formData.append(item, queryString[item]);
    });

    const response = await fetch(this.config.token_url, {
      method: 'post',
      headers: new Headers({
        Accept: 'application/json',
      }),
      body: formData,
    });

    if (response.ok) {
      const json = await response.json();

      if (json.error) {
        this.processing = false;
        throw new Error(json.error.message);
      }

      this.processing = false;
      return this.storeToken(json);
    }
  }

  /**
   * Exchange your refresh token for a new auth token.
   *
   * @param token
   * @param scope
   *
   * @returns {Promise<void>}
   *  Returns a Promise that resolves with the new token retrieved from Drupal.
   */
  // eslint-disable-next-line consistent-return
  async refreshOauthToken(token) {
    if (token !== null) {
      const formData = new FormData();
      formData.append('grant_type', 'refresh_token');
      formData.append('client_id', this.config.client_id);
      formData.append('client_secret', this.config.client_secret);
      formData.append('refresh_token', token.refresh_token);

      const response = await fetch(this.config.token_url, {
        method: 'post',
        headers: new Headers({
          Accept: 'application/json',
          'X-Consumer-ID': `${this.config.client_id}`,
        }),
        body: formData,
      });

      if (response.ok) {
        const json = await response.json();

        if (json.error) {
          throw new Error(json.error.message);
        }

        return this.storeToken(json);
      }

      throw new Error(response.status);
    }
  }

  // eslint-disable-next-line class-methods-use-this
  storeToken(json) {
    const token = Object.assign({}, json);
    token.date = Math.floor(Date.now() / 1000);
    token.expirationDate = token.date + token.expires_in;
    localStorage.setItem('drupal-oauth-token', JSON.stringify(token));
    return token;
  }
}

export default DrupalOauth;
