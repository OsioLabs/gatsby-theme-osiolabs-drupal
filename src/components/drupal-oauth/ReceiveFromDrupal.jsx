import React from 'react';
import PropTypes from 'prop-types';
import qs from 'qs';
import {OauthReceiver} from 'react-oauth-flow';
import withCurrentLocation from '../withCurrentLocation';
import withDrupalOauthConsumer from './withDrupalOauthConsumer';

class ReceiveFromDrupal extends React.Component {
  state = {
    codePresent: false,
  };

  componentDidMount() {
    if (/code=/.test(window.location.search)) {
      this.setState({ codePresent: true });
    }
  }

  handleSuccess = (accessToken, { response, state }) => {
    const { updateAuthenticatedUserState } = this.props;
    updateAuthenticatedUserState(true);
  };

  handleError = error => {
    // @todo: What should we do when this doesn't work?
  };

  // Drupal's Simple Oauth module expects parameters to be be form encoded and
  // in the body of the request. The default fetch function of the OauthReceiver
  // component uses query string arguments. So we provide our own implementation.
  fetchToken = (url, fetchArgs) => {
    try {
      const token = this.props.drupalOauthClient.fetchOauthToken(url)
        .then(response => {
          return response;
        });
      return token;
    } catch (e) {
      throw e;
    }
  };

  render() {
    // Normalize the redirect URI so that it always contains a trailing slash.
    // Gatsby likes to redirect example.com/asdf?code= to example.com/asdf/?code=
    // and if the URL that we send in the original request doesn't match the one
    // we use to verify the ?code= param there will be problems.
    const {
      currentLocation,
      currentLocationRoot,
      drupalOauthClient,
    } = this.props;

    const { codePresent } = this.state;

    let redirectUri = `${currentLocationRoot}${currentLocation.pathname}`;

    redirectUri += redirectUri.endsWith('/') ? '' : '/';

    if (codePresent) {
      return (
        <OauthReceiver
          tokenUrl={drupalOauthClient.config.token_url}
          clientId={drupalOauthClient.config.client_id}
          clientSecret={drupalOauthClient.config.client_secret}
          redirectUri={redirectUri}
          onAuthSuccess={this.handleSuccess}
          onAuthError={this.handleError}
          tokenFn={this.fetchToken}
          render={({ processing, state, error }) => (
            <div data-testid="receiver" style={{display: 'none'}}>
              {processing && <p>Authorizing now...</p>}
              {error && (
                <p className="error">An error occurred: {error.message}</p>
              )}
            </div>
          )}
        />
      );
    }

    // Render nothing.
    return <></>;
  }
}

ReceiveFromDrupal.propTypes = {
  drupalOauthClient: PropTypes.object.isRequired,
  updateAuthenticatedUserState: PropTypes.func.isRequired,
  currentLocation: PropTypes.object.isRequired,
  currentLocationRoot: PropTypes.string.isRequired,
};

export default withCurrentLocation(withDrupalOauthConsumer(ReceiveFromDrupal));
