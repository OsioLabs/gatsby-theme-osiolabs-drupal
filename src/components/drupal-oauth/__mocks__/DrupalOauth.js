// __mocks__/DrupalOauth.js

const mock = jest.fn().mockImplementation(() => ({
  config: {
    drupal_root: 'http://api.example.com',
    authorize_url: 'http://api.example.com/oauth/authorize',
    token_url: 'http://api.example.com/oauth/token',
    client_id: 'CLIENTID',
    client_secret: 'CLIENTSECRET',
  },
  isLoggedIn: jest.fn(() =>
    Promise.resolve({ access_token: 'access-token', token_type: 'Bearer' })
  ),
  handleLogin: jest.fn(() => Promise.resolve({})),
  handleLogout: jest.fn(() => Promise.resolve({})),
  fetchOauthToken: jest.fn(() => Promise.resolve({})),
  storeToken: jest.fn(),
}));

mock.createFromEnvironment = jest.fn().mockImplementation(mock);

export default mock;
