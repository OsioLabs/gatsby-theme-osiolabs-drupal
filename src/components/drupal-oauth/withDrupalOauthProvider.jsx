import React from 'react';
import DrupalOauthContext from './DrupalOauthContext';

/**
 * HOC to assist in initializing a Drupal OAuth context provider.
 *
 * Creates a new context with the following values:
 * - userAuthenticated: A boolean indicating if the current user is
 *   authenticated.
 * - drupalOauthClient: An instance of the DrupalOauth client from
 *   drupalOauthClient.js which can be used to access information about the
 *   current user.
 * - updateAuthenicaticatedUserState: A function that can be called by child
 *   components to update the value of userAuthenticated within the existing
 *   context.
 *
 * @param {*} client
 *   And instance of the drupalOAuth class from drupalOAuth.js that has already
 *   been initialized.
 * @param {*} Component
 *   The child component to wrap with the new context provider.
 */
const withDrupalOauthProvider = (client, Component) => {
  class WithDrupalOauthProvider extends React.Component {
    constructor(props) {
      super(props);

      this.state = {
        userAuthenticated: false,
        drupalOauthClient: client,
        updateAuthenticatedUserState: newState => {
          this.setState({ userAuthenticated: newState });
        },
      };

      this.state.drupalOauthClient
        .isLoggedIn(this.state.updateAuthenticatedUserState)
        .then(token => {
          if (token !== false) {
            this.setState({ userAuthenticated: true });
          }
        })
        .catch(error => {
          this.setState({ userAuthenticated: false });
        });
    }

    render() {
      return (
        <DrupalOauthContext.Provider value={this.state}>
          <Component {...this.props} />
        </DrupalOauthContext.Provider>
      );
    }
  }

  return WithDrupalOauthProvider;
};

export default withDrupalOauthProvider;
