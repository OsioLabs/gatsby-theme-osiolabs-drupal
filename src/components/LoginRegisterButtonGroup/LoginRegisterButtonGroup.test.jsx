import React from 'react';
import { render, cleanup } from '@testing-library/react';

// Configure some mocks we need to make this component work.
import DrupalOauth from '../drupal-oauth/DrupalOauth';
import withDrupalOauthProvider from '../drupal-oauth/withDrupalOauthProvider';

import LoginRegisterButtonGroup from './LoginRegisterButtonGroup';
import ReceiveFromDrupal from '../drupal-oauth/ReceiveFromDrupal';

jest.mock('../drupal-oauth/DrupalOauth');
const drupalOauthClient = new DrupalOauth();
ReceiveFromDrupal.fetchToken = jest.fn(() => Promise.resolve({}));

const TestElement = withDrupalOauthProvider(drupalOauthClient, () => (
  <LoginRegisterButtonGroup />
));

describe('Component: <LoginRegisterButtonGroup />', () => {
  beforeEach(() => {
    fetch.resetMocks();
    drupalOauthClient.storeToken.mockClear();
  });

  afterEach(cleanup);

  // Test the the component renders correctly with a simple snapshot test.
  it('renders correctly', () => {
    // Basic snapshot test.
    const { asFragment } = render(<TestElement />);
    const firstRender = asFragment();
    expect(firstRender).toMatchSnapshot();
  });
});
