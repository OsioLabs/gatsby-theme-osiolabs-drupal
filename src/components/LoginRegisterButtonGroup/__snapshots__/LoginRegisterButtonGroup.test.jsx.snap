// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <LoginRegisterButtonGroup /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="ui buttons"
  >
    <a
      class="ui button"
      href="http://api.example.com/oauth/authorize?client_id=CLIENTID&redirect_uri=http%3A%2F%2Flocalhost%2F&response_type=code&state=%7B%7D"
    >
      Log in
    </a>
    <div
      class="or"
      data-text="/"
    />
    <a
      class="ui button primary"
      href="http://api.example.com/user/register?destination=/oauth/authorize%3Fclient_id%3DCLIENTID%26redirect_uri%3Dhttp%253A%252F%252Flocalhost%252F%26response_type%3Dcode%26state%3D%257B%257D"
    >
      Sign up
    </a>
  </div>
</DocumentFragment>
`;
